benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,944 @@
+                record_event(
+                    self.rank,
+                    GRADIENT_SERVER_BATCH_METRIC,
+                    self.average_gradient_bp_key(
+                        self.batch_number
+                    ),
+                    BP_LOC_BATCH,
+                    self.metrics
+                )
+                if self.batch_mode:
+                    self.gradient[bp_loc] = [gradient]
+                else:
+                    self.gradient[bp_loc] = gradient
+                self.futures[bp_loc] = [fut]
+            else:
+                if self.batch_mode:
+                    self.gradient[bp_loc].append(gradient)
+                else:
+                    self.gradient[bp_loc] += gradient
+                self.futures[bp_loc].append(fut)
+            if len(self.futures[bp_loc]) == self.trainer_count:
+                record_event(
+                    self.rank,
+                    GRADIENT_SERVER_STRAGGLER_METRIC,
+                    self.average_gradient_bp_key(
+                        self.batch_number
+                    ),
+                    BP_LOC_STRAGGLER,
+                    self.metrics
+                )
+                if self.batch_mode:
+                    # TODO: cuda kernel
+                    bp_loc_avg = self.gradient[bp_loc][0]
+                    for i in range(1, self.trainer_count):
+                        bp_loc_avg += self.gradient[bp_loc][i]
+                else:
+                    bp_loc_avg = self.gradient[bp_loc]
+                bp_loc_avg / (1.0 * self.trainer_count)
+                if not self.use_cuda_rpc:
+                    bp_loc_avg = bp_loc_avg.cpu()
+                if sparse_gradient:
+                    if self.backend == GLOO:
+                        bp_loc_avg = bp_loc_avg.coalesce()
+                    bp_loc_avg = [
+                        bp_loc_avg._indices(),
+                        bp_loc_avg._values(),
+                        bp_loc_avg.size()
+                    ]
+                for cur_fut in self.futures[bp_loc]:
+                    cur_fut.set_result(bp_loc_avg)
+                record_event(
+                    self.rank,
+                    GRADIENT_SERVER_BATCH_METRIC,
+                    self.average_gradient_bp_key(
+                        self.batch_number
+                    ),
+                    BP_LOC_BATCH,
+                    self.metrics
+                )
+                return fut
+        return fut
+
+    def rpc_warmup_call(tensor):
+        return tensor
+
+    def reset(gs_rref):
+        self = gs_rref.local_value()
+        self.clear_batch_state()
+        self.batch_number = 0
+
+    def get_metrics(gs_rref):
+        self = gs_rref.local_value()
+        torch.cuda.synchronize(self.rank)
+        return [self.rank, process_metrics_for_elapsed_time(self.metrics)]
+
+
+class HookState():
+    def __init__(
+        self,
+        rank,
+        process_group,
+        process_group_size,
```

Comment:
can this be retrieved from `process_group.size()`?

Reply:
yes I can change it.